src/main/java/org/dataone/speedbagit/SpeedBagIt.java

```diff
@@ -119,98 +178,108 @@ public void writeToTagManifest(String path, String checksum) {
         }
     }
 
+    private void streamFile(ZipOutputStream zos, SpeedFile streamingFile) throws IOException {
+        try {
+            ZipEntry entry = new ZipEntry(streamingFile.getPath());
+            zos.putNextEntry(entry);
+
+            SpeedStream fileStream = streamingFile.getStream();
+            fileStream.transferTo(zos);
+        } finally {
+            zos.closeEntry();
+        }
+    }
 
     /**
+     * Writes the files to a stream under the BagIt specification. The manifest, bagit.txt,
+     * and bag.info are generated inside.
+     *
      * @param zos: The output stream that represents the streaming bag
-     * @throws IOException
+     * @throws IOException Throws when something went wrong with streaming the bag
+     * @throws NoSuchAlgorithmException Thrown when an unsupported checksum algorithm is used
      */
     public void stream(ZipOutputStream zos)
             throws IOException, NoSuchAlgorithmException {
-
-        logger.info("stream called!");
+        int totalSize = 0;
+        // Stream all of the files in the root 'data' directory
         for (SpeedFile streamingFile : this.dataFiles) {
-            logger.info(String.format("Preparing to write %s", streamingFile.getPath()));
-            ZipEntry entry = new ZipEntry(streamingFile.getPath());
-            zos.putNextEntry(entry);
-
-            SpeedStream fileStream = streamingFile.getStream();
-            fileStream.transferTo(zos);
+            this.streamFile(zos, streamingFile);
 
             String checksum = new String(streamingFile.getStream().getChecksum());
-            logger.info(checksum);
-            logger.info(String.valueOf(streamingFile.getStream().getSize()));
             this.writeToDataManifest(streamingFile.getPath(), checksum);
-            this.payloadOxum += streamingFile.getStream().getSize();
-            zos.closeEntry();
+            totalSize += streamingFile.getStream().getSize();
         }
+        String payloadOxum =  String.format("%s.%s",totalSize, this.dataFiles.size());
+        // Generate and add the bagit.txt file
+        InputStream bagTextStream = new ByteArrayInputStream(this.generateBagitTxt().getBytes(StandardCharsets.UTF_8));
+        this.addFile(bagTextStream, "bagit.txt", MessageDigest.getInstance(checksumAlgorithm), true);
 
-        // Now that payload-oxum is computed, add it to the bag-info.txt file
-        this.bagInfoFile = String.format("%spayload-oxum: %s\n", this.bagInfoFile, payloadOxum);
-        InputStream fileStream = new ByteArrayInputStream(this.bagInfoFile.getBytes(StandardCharsets.UTF_8));
-        this.addFile(fileStream, "bag-info.txt", MessageDigest.getInstance("MD5"), true);
+
+        // Generate and add the bag-info.txt file
+        String bagInfoFile = generateBagInfoTxt(payloadOxum, totalSize);
+        InputStream fileStream = new ByteArrayInputStream(bagInfoFile.getBytes(StandardCharsets.UTF_8));
```

Comment:
Can the bag info file get huge for huge bags? If so, `ByteArrayInputStream` may be affected to in-memory limits, but if even the biggest files result in only Megabyte bag info files, then it's likely not a big deal.

Reply:
The bag-info.txt file size O(n) where n is the number of metadata fields a user puts in the file. BagIt only reserves ten tags that are optional, [here](https://www.rfc-editor.org/rfc/rfc8493.html#section-2.2.2). I don't think that this should be an issue.